contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -3,16 +3,117 @@
+      return eosio::sha256(finalizer_key_binary.data(), finalizer_key_binary.size());
+   }
+
+   // Returns hash of finalizer_key in text format
    static eosio::checksum256 get_finalizer_key_hash(const std::string& finalizer_key) {
-      const auto fin_key_g1 = eosio::decode_bls_public_key_to_g1(finalizer_key);
-      return eosio::sha256(fin_key_g1.data(), fin_key_g1.size());
+      const auto fin_key_g1 = to_binary(finalizer_key);
+      return get_finalizer_key_hash(fin_key_g1);
+   }
+
+   finalizers_table::const_iterator system_contract::get_finalizer_itr( const name& finalizer_name ) const {
+      // Check finalizer has registered keys
+      auto finalizer_itr = _finalizers.find(finalizer_name.value);
+      check( finalizer_itr != _finalizers.end(), "finalizer " + finalizer_name.to_string() + " has not registered any finalizer keys" );
+      check( finalizer_itr->finalizer_key_count > 0, "finalizer " + finalizer_name.to_string() + "  must have at least one registered finalizer keys, has " + std::to_string(finalizer_itr->finalizer_key_count) );
+
+      return finalizer_itr;
+   }
+
+   // This function may never fail, as it can be called by update_elected_producers,
+   // and in turn by onblock
+   // Establish finalizer policy from `proposed_fin_keys` and calls
+   // eosio::set_finalizers host function
+   void system_contract::set_proposed_finalizers( std::vector<finalizer_auth_info>& proposed_finalizers ) {
+      std::sort( proposed_finalizers.begin(), proposed_finalizers.end(), []( const finalizer_auth_info& lhs, const finalizer_auth_info& rhs ) {
+         return lhs.key_id < rhs.key_id;
+      } );
+
+      const auto last_proposed_finalizers = get_last_proposed_finalizers();
+      if( proposed_finalizers == last_proposed_finalizers ) {
+         // Finalizer policy has not changed. Do not proceed.
+         return;
+      }
+
+      // Construct finalizer authorities
+      std::vector<eosio::finalizer_authority> finalizer_authorities;
+      finalizer_authorities.reserve(proposed_finalizers.size());
+      for( const auto& k: proposed_finalizers ) {
+         finalizer_authorities.emplace_back(k.fin_authority);
+      }
+
+      // Establish finalizer policy
+      eosio::finalizer_policy fin_policy {
+         .threshold  = ( finalizer_authorities.size() * 2 ) / 3 + 1,
+         .finalizers = std::move(finalizer_authorities)
+      };
+
+      // call host function
+      eosio::set_finalizers(std::move(fin_policy)); // call host function
+
+      // store last proposed policy in both cache and DB table
+      _last_prop_finalizers_cached = proposed_finalizers;
+      if( _last_prop_finalizers.begin() == _last_prop_finalizers.end() ) {
+         _last_prop_finalizers.emplace( get_self(), [&]( auto& f ) {
+            f.last_proposed_finalizers = proposed_finalizers;
+         });
+      } else {
+         _last_prop_finalizers.modify(_last_prop_finalizers.begin(), same_payer, [&]( auto& f ) {
```

Comment:
Once again, storing `_last_prop_finalizers.begin()` as an iterator allows avoid duplicate calls to `begin()` which I believe are expensive because they require doing a lower_bound database lookup.